### ISSUE TITLE
FLAC: Mismatch with max delta from 0.085 to 1.998

### ISSUE THREAD

**Issue by totikom** [pdeljanov/Symphonia#74]
[These files](https://keybase.pub/totikom/Symphony%20failed%20tests/) exhibit mismatches compared to ffmpeg:
> 1.25170898 01 - Give It Away (In Progress).flac
 1.93112183 05 - Киберпанк.flac
 1.59304810 06. The Sunshine.flac
 1.98730469 03 Roll Us A Giant.flac
 1.72723389 10. The Parts.flac
 1.08828735 07. The Grocery.flac
 0.08523428 04- The Hours - Arr. Michael Riesman - Morning Passages.flac
 1.66995239 16. Olivia's Doom (Chad Mossholder Remix).flac
 1.92807007 09 - Путь прост.flac
 1.92960131 40. Stumpy.flac
 1.99769855 3-Blue Jeans.flac
 1.92474365 Robert Rich - Echo of Small Things.flac

More over, [one](https://keybase.pub/totikom/Symphony%20failed%20tests/01.%20You%20Are%20Light%20(feat.%20Felicia%20Farerre).flac) of the files triggers an error:
> ERROR symphonia_core::probe > reached probe limit of 1048576 bytes.Test interrupted by error: unsupported feature: core (probe): no suitable reader found 01. You Are Light (feat. Felicia Farerre).flac

**Comment by totikom** [comment]
Tested on commit 0b67a2ddc3b4600f51f5c94d7551a70fffb5ca56

**Comment by pdeljanov** [comment]
Hi @totikom,

Thanks for testing. The majority of the files should now be fixed with the latest commit.

I'm curious, how many files did you have to check in total to find these?

---

```
06. The Sunshine.flac
10. The Parts.flac
07. The Grocery.flac
Robert Rich - Echo of Small Things.flac
```

The following files are damaged and FFMpeg chooses to drop more audio packets than Symphonia. This causes the decoded output of FFMPEG and Symphonia to become "mis-aligned" and then every sample after that mismatches.

I don't consider this to be a bug because FLAC packets are independent of each other, and are checksummed. Therefore, preserving as many uncorrupted packets as possible is optimal in my opinion.

```
40. Stumpy.flac
```

This file contains a packet with a checksum error. Symphonia drops it, but FFMpeg preserves it. This causes a similar "mis-alignment" issue as noted above.

I don't consider this a bug either because Symphonia can't prove if the CRC is simply wrong or if the packet is corrupted. It is feasible for a corrupt packet to be successfully decoded but also produce noise at the output. This is something I want to avoid.

```
01. You Are Light (feat. Felicia Farerre).flac
```

This file is only filled with 0s and not playable in general. :laughing:

**Comment by Shnatsel** [comment]
> I'm curious, how many files did you have to check in total to find these?

I don't have an exact number of files, but they take up at around 500Gb.

**Comment by totikom** [comment]
>I'm curious, how many files did you have to check in total to find these?

Number of files would not be representative, as there are lots of full album images,  but the overall duration of that files  is more than a month.

**Comment by Nathy-bajo** [comment]
I am also having issues on this when I try to play the audio samples I am getting from my client.
This is my playback thread.

```rust
let play_back_thread = tokio::spawn(async move {
        //  create the rodio output stream
        println!("castle");
        let (_, stream_handle) = OutputStream::try_default().unwrap();
        let spec = hound::WavSpec {
            channels: 1,
            sample_rate: 44100,
            bits_per_sample: 16,
            sample_format: hound::SampleFormat::Int,
        };
        println!("spec: {:?}", spec);

        while let Some(packet) = playback_recv.recv().await {
            let mut buffer = std::io::Cursor::new(Vec::new());
            println!("killstreak");

            let mut writer = hound::WavWriter::new(&mut buffer, spec.clone()).unwrap();
            for sample in packet.data {
                writer.write_sample(sample as i16).unwrap();
            }
            writer.finalize().unwrap();
            let source = rodio::Decoder::new(buffer).unwrap();
            println!("I am");
            stream_handle.play_raw(source.convert_samples()).unwrap();
        }
    });
    ```
    The error I am getting.
    ```
[2022-11-06T09:28:50Z ERROR symphonia_core::probe] reached probe limit of 1048576 bytes.
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: UnrecognizedFormat'
```
or am I missing something please?

**Comment by Nathy-bajo** [comment]
I am also facing this issue when trying to play audio samples that I am getting from my client.
This is my playback thread.

```rust
    let play_back_thread = tokio::spawn(async move {
        //  create the rodio output stream
        let (_, stream_handle) = OutputStream::try_default().unwrap();
        let spec = hound::WavSpec {
            channels: 1,
            sample_rate: 44100,
            bits_per_sample: 16,
            sample_format: hound::SampleFormat::Int,
        };

        while let Some(packet) = playback_recv.recv().await {
            let mut buffer = std::io::Cursor::new(Vec::new());

            let mut writer = hound::WavWriter::new(&mut buffer, spec.clone()).unwrap();
            for sample in packet.data {
                writer.write_sample(sample as i16).unwrap();
            }
            writer.finalize().unwrap();
            let source = rodio::Decoder::new(buffer).unwrap();
            stream_handle.play_raw(source.convert_samples()).unwrap();
        }
    });
    ```
`[2022-11-06T09:28:50Z ERROR symphonia_core::probe] reached probe limit of 1048576 bytes.
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: UnrecognizedFormat'
 `
 
    
 And this is the error I am getting. Or am I doing something wrong please?

**Comment by pdeljanov** [comment]
@Nathy-bajo, that error means Symphonia couldn't detect the type of audio file it was given. I'd recommend making sure the file is an audio file, and supported by Symphonia.

Since you are using Rodio, make sure you also enable all Symphonia decoders with Rodio.

If you are pretty certain the file should play, please provide the output of this FFMpeg command:

`ffprobe -show_streams -show_format -of json </path/to/the/broken/file>`.

Cheers!

**Comment by Nathy-bajo** [comment]
> @Nathy-bajo, that error means Symphonia couldn't detect the type of audio file it was given. I'd recommend making sure the file is an audio file, and supported by Symphonia.
> 
> Since you are using Rodio, make sure you also enable all Symphonia decoders with Rodio.
> 
> If you are pretty certain the file should play, please provide the output of this FFMpeg command:
> 
> `ffprobe -show_streams -show_format -of json </path/to/the/broken/file>`.
> 
> Cheers!

I am trying to play a real-time audio buffer that is coming from my client not a file. It is a stream of PCM 16bit data from my client device microphone

**Comment by totikom** [comment]
Btw, @pdeljanov  I'm still have tons of audiofiles, and I am willing to use them for testing Symphonia. Feel free to call me before releases for stress testing them.

**Comment by Nathy-bajo** [comment]
> Symphonia can't detect raw PCM because there is no container format. You'll need to use `PcmDecoder` directly. I don't think this is possible with the built-in Symphonia decoder in Rodio.
> 
> If you use Symphonia directly you can do something like this:
> 
> ```rust
> /* Setup Decoder */
> let mut params = CodecParameters::new();
> 
> params.for_codec(CODEC_TYPE_PCM_S16LE)
>     .with_channels(/* number of channels */)
>     .with_max_frames_per_packet(/* number of audio frames in a packet */)
>     .with_bits_per_coded_sample(16)
>     .with_bits_per_sample(16)
>     .with_sample_rate(/* sample rate */)
>     .with_time_base(TimeBase::new(1, /* sample rate */));
> 
> let mut decoder = symphonia::default::get_codecs().make(&params, &Default::default())?;
> 
> 
> /* Playback loop */
> 
> let packet = Packet::new_from_slice(0, /* timestamp (optional) */, /* duration (optional) */, /* audio data slice */)
> 
> let audio = decoder.decode(&packet)?;
> ```
> 
> Since you are just getting raw PCM, you may not even need Symphonia at all.

I don't understand what you mean by I won't need `symphonia` as I am still new to pcm audio aspect.
I tried rodio it did not work so I stumbled upon another way I could play my stream of PCM 16bit data from my client device microphone.
https://github.com/mnkrana/record-audio/blob/e3effa46bc68e65d31a3a5fb8ed33c11bb4d173a/src/audio_clip.rs#L128
They are playing from a file but I am playing from a buffer that is holding audio data. Although it works and plays the the PCM stream but it was not what was expect. The sound it was playing is like a radio which is not tuned well.
The client sends 16bit pcm data but the player only plays f64 bits which is probably why I am not getting the actual sound I want to get. But maybe converting the incoming `Int16ToFloat32` might do the trick.
I don't know if It is that I am wrong somewhere and not doing something right. So I would really appreciate if you could explain a bit more on how I am supposed to play the stream.

(Thanks for porting this library to rust btw :heart: )

**Comment by pdeljanov** [comment]
Hi @Nathy-bajo 

PCM audio is basically the raw, un-compressed/encoded audio. Symphonia takes compressed/encoded audio and produces buffers of decoded PCM audio. The PCM audio buffer can then be passed to the OS for playback. However, since you already have PCM audio buffers, there's no need to use Symphonia.

There are a few important parameters for a PCM audio buffer:
1. The sample format (e.g., i16, i32, f64, etc.). This is the data type of each audio sample in the buffer.
2. The number of audio channels.
3. The sample rate. How many samples are need per 1 second of audio.

Usually, your PCM audio buffer will store audio samples in an interleaved order. Basically, 1 sample per channel, before the next sample.

For example, if you have stereo audio (left and right channels), your buffer will contain samples in this order: L, R, L, R, L, R. The pairs of samples are called a frame.

Therefore, if your audio buffer has 50 frames of 2 channel audio, the buffer contains 100 samples.

The sample rate isn't too important for the buffer itself, but you need to know the sample rate to configure the output audio device.

Knowing this, if your audio sounds off, there could be a number of reasons:
1. You didn't configure the number of channels on the output device to match the number of channels in the audio buffer. You need to configure the device with the correct number of channels.
2. The sample format of the audio buffer doesn't match that of the output device. You'll need to convert it to match.
3. The sample rate configured on the output device doesn't match the audio buffer. You need to configure the device with the correct sample rate.

Hope this is clear and helps!

---

Hi @totikom,

Thanks! I've been fixing issues with the AAC decoder. We've test around 5000 files, but a larger corpus would be great. Do you have any?

**Comment by Nathy-bajo** [comment]
> Hi @Nathy-bajo
> 
> PCM audio is basically the raw, un-compressed/encoded audio. Symphonia takes compressed/encoded audio and produces buffers of decoded PCM audio. The PCM audio buffer can then be passed to the OS for playback. However, since you already have PCM audio buffers, there's no need to use Symphonia.
> 
> There are a few important parameters for a PCM audio buffer:
> 
> 1. The sample format (e.g., i16, i32, f64, etc.). This is the data type of each audio sample in the buffer.
> 2. The number of audio channels.
> 3. The sample rate. How many samples are need per 1 second of audio.
> 
> Usually, your PCM audio buffer will store audio samples in an interleaved order. Basically, 1 sample per channel, before the next sample.
> 
> For example, if you have stereo audio (left and right channels), your buffer will contain samples in this order: L, R, L, R, L, R. The pairs of samples are called a frame.
> 
> Therefore, if your audio buffer has 50 frames of 2 channel audio, the buffer contains 100 samples.
> 
> The sample rate isn't too important for the buffer itself, but you need to know the sample rate to configure the output audio device.
> 
> Knowing this, if your audio sounds off, there could be a number of reasons:
> 
> 1. You didn't configure the number of channels on the output device to match the number of channels in the audio buffer. You need to configure the device with the correct number of channels.
> 2. The sample format of the audio buffer doesn't match that of the output device. You'll need to convert it to match.
> 3. The sample rate configured on the output device doesn't match the audio buffer. You need to configure the device with the correct sample rate.
> 
> Hope this is clear and helps!
> 
> Hi @totikom,
> 
> Thanks! I've been fixing issues with the AAC decoder. We've test around 5000 files, but a larger corpus would be great. Do you have any?

Thank you very much for this clear explanation I really needed this.

**Comment by totikom** [comment]
> Hi @totikom,
> 
> Thanks! I've been fixing issues with the AAC decoder. We've test around 5000 files, but a larger corpus would be great. Do you have any?

Unfortunately, no. I'm kinda lossless fan, so I have lots of FLACs and APEs (even some SACDs and DXDs, if you want), but no AACs and practically no MP3s...

**Comment by pdeljanov** [comment]
> > Hi @totikom,
> > Thanks! I've been fixing issues with the AAC decoder. We've test around 5000 files, but a larger corpus would be great. Do you have any?
> 
> Unfortunately, no. I'm kinda lossless fan, so I have lots of FLACs and APEs (even some SACDs and DXDs, if you want), but no AACs and practically no MP3s...

Got it! I'm the same way apart from the old parts of my collection. :) 

There haven't been any changes to FLAC for a while, but I think in a future release I'll be reworking the FLAC demuxer. I'll be sure to rope you in for testing when that changes happens.

**Comment by pdeljanov** [comment]
Hey @totikom,

Okay, you got me interested!

I rewrote the FLAC packet parser in the `flac-parser-v3` branch. The branch contains all the changes I wanted to make to the FLAC demuxer.

The packet parser is used by the demuxer to take the `MediaSourceStream` and slice it into packets. This is actually pretty difficult with FLAC because FLAC packets don't have a stated length so generally you need to decode a packet to know where it ends.

Instead of decoding a packet to find its length (inefficient), Symphonia tries to figure out the packet boundaries using the packet checksum and other information.

The existing implementation wasn't great, complicated, and ran into issues with some files. On my music collection, 102 files out 3042 failed verification due to packets being missed or dropped by the packet parser.

I've tried to address all the deficiencies in the new branch. With the new packet parser 0 files fail verification.

Could you please test it against your collection?

Please use `RUST_LOG=info symphonia-play --verify-only /path/to/file.flac` since that compares the decoded audio against the internal MD5 checksum of the raw audio.

I use the following command to automate this:

```bash
find ~/Music -type f -name \*.flac -exec echo {} \; -exec ./target/release/symphonia-play --verify-only {} \; > test.log 2>&1
```

If a file fails, please check if the reference decoder also fails to verify file: `flac -t /path/to/file.flac`.

If you are bored, it'd also be interesting to see how it compares against the current mainline code.

@Shnatsel, please give it a try if you can too.

Thanks!

**Comment by Shnatsel** [comment]
I'll check the FLAC I have lying around.

Here's a torrent with 750GB of freely redistributable FLAC - it's the Pony Music Archive:
```
magnet:?xt=urn:btih:a28d9ff9d8e1edc220b4511c182345d71739b8c7&dn=Pony%20Music%20Archive%2021.03%20%28Raw%20Quality%29&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce
```

I've used it for testing before, but I don't have the necessary 750GB of disk space to download it now. If that's too much for you as well, let me know and I'll figure something out.

**Comment by Shnatsel** [comment]
`symphonia-play` is not very friendly to automation on account of always printing 'error: end of stream' and exiting with code 1, but I have written a script that detects and reports actual failures:

```bash
#!/bin/sh

output=$(target/release/symphonia-play --verify-only 2>&1 "$1" | grep -v 'ERROR symphonia_play > error: end of stream')

if (echo "$output" | grep -q 'ERROR' ) || (echo "$output" | grep -q 'panicked' ); then
    echo "$1"
    echo "$output"
fi
```

Save it to a file and name it `symphonia-verify.sh`

You can then run it **in parallel (using all CPU cores for decoding)** like this: 

`find ~/Music -type f -name \*.flac | nice ionice parallel symphonia-verify.sh`

Make sure you have GNU parallel installed. 

`nice ionice` is to give the process low CPU and disk priority, they're not actually required.

**Comment by pdeljanov** [comment]
Yeah, using `symphonia-play` for verification testing is crude.

To clarify though, the output in question is a debug level log stating "verification failed". A successful verification will print a debug log of "verification passed".

I just piped the output from the command I provided to a file and searched in VS Code for the number of occurrences of those phrases lol...

That torrent is more than double the amount of storage I have, but I may pick some random files from it. Most FLAC files are encoded via. libflac so they're more consistent than something like MP3 or AAC.

**Comment by Shnatsel** [comment]
I'm not seeing the passed/failed messages. Did you forget to include a LOG_LEVEL variable into the command?

I'd argue that `--verify-only` should print the result without requiring extra manual steps.

**Comment by pdeljanov** [comment]
Agreed, just working with what we have right now.

I edited the command, forgot that I had `RUST_LOG` exported.

**Comment by Shnatsel** [comment]
I've updated the script for `RUST_LOG`.

**Comment by Shnatsel** [comment]
Okay, no issues on the files I had lying around. I'll download the torrent myself then, I think I know how to do that.

If you run into issues like these in the future, public clouds are a good solution for large short-lived compute jobs. Google Cloud gives everyone $300 in free credit for 3 months, Linode gives people $100 free credit.

**Comment by roderickvd** [comment]
Amazing work guys. I am working on a new streaming client that will use MP3 and FLAC (can you guess which? 😄) and this is just awesome.

If you would ever consider it, I would be happy to chip in a donation towards buying computing time or an external hard drive.

**Comment by totikom** [comment]
I've ran Shnatsel's script and it completed quite fast without _any_ errors, which sounded **suspicious** (don't get me wrong, I have an old drive, so bit rod occurs from time to time). So i've started the original [command](https://github.com/pdeljanov/Symphonia/issues/74#issuecomment-1312859520).
Well, it is still running and logs already have a bunch of failures.

I'll recheck them with `flac` when `symphonia-play` finishes and report the results.

**Comment by Shnatsel** [comment]
I've just checked and the script does work for me on separate files. It does correctly report failures. So I'm very confused.

What does `parallel --version` say? Maybe it's not GNU parallel.

(it does print more than it should in some cases, that's a bug. but I don't want to edit the version posted above to complicate debugging)

**Comment by Shnatsel** [comment]
Oh, I know why it didn't work :facepalm: 

I have left the absolute path to the script on *my* system in it, and the script didn't complain that it couldn't find the file. Here's a version that uses a relative path and does complain if it can't find it:

```bash
#!/bin/sh

set -e
binary='target/release/symphonia-play'
"$binary" --help > /dev/null

output=$(RUST_LOG=debug "$binary" --verify-only 2>&1 "$1" | grep -v 'error: end of stream')

if (echo "$output" | grep -q 'ERROR' ) || (echo "$output" | grep -q 'verification failed' ) || (echo "$output" | grep -q 'panicked' ); then
    echo "$1"
    echo "$output"
fi
```

**Comment by totikom** [comment]
Of course, I've changed the path, but looks like I've mistyped some args in`find`:facepalm:.

Anyway, I already have the needed data, I just didn't have time to check it for false-negatives.

Hopefully, I'll do it in a few hours.

**Comment by Shnatsel** [comment]
I've found 6 files that fail validation in Symphonia but pass `flac -t`, [here's the archive](https://mega.nz/file/9skQGJyS#HSB-yO0nlT2gN3agegjteT-jKtloahS4UQNCT7qojyQ).

All of these come from the Pony Music Archive corpus.

**Comment by pdeljanov** [comment]
Got it!

Version 1.4 was a major update to the reference decoder/encoder. The issue with those particular files are due to an invalid bitstream. I guess the previous versions didn't do enough bitstream validation. FFMpeg also seems to reject the packet in question so I think we're good re: the Pony archive.

**Comment by totikom** [comment]
Sounds like all files, that failed `symphonia-play` also failed `flac -t`.
All with `FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC` so it seems, that anything is good.

P.S. I'm using `flac 1.4.2`

P.P.S I should write a script, that will pass an audio file to `flac` and, if it is accepted, pass it further to `symphonia` binary under test. That way we will be automatically assured, that all the fails are `symphonia`'s divergence from `flac` behavior.

**Comment by pdeljanov** [comment]
Thanks @totikom, sounds like the new parser is strictly an improvement so I think I'll merge it shortly.

When I have a moment I'll update `symphonia-play` to make the output of verification mode more useful and return a proper error code. That should help scripting this in the future.

Just a heads up, `symphonia-check --ref flac` will compare the output of the FLAC reference decoder against Symphonia's output. However, this isn't the same as checking against the file's MD5. Nor is it the same as running `flac -t`. But maybe this functionality should to be added to `symphonia-check` instead of a separate script.

**Comment by pdeljanov** [comment]
Merged in https://github.com/pdeljanov/Symphonia/commit/f21afd5ed5e8e10c9db49a69f98a9b6b033f1d7c!

If there are any future issues with FLAC, let's open a new issue. 

Thanks for the help testing!

**Comment by Nathy-bajo** [comment]
Hi @pdeljanov 

So sorry for disturbing you again. You said since I am sending raw pcm 16bit data I don't need to decode it using symphonia so I tried playing it directly and the output was like a noisy siren.

Currently I am using `rust-portaudio` library for sending audio buffers to the client and receiving from the client.
This is my code: 

```rust
const FRAMES: u32 = 256;
const INTERLEAVED: bool = true;

async fn main() -> Result<(), pa::Error> {

    let (send_audio, recv_audio) = tokio::sync::watch::channel(Vec::new());

    let audio_thread = std::thread::spawn(move || {
        let pa = pa::PortAudio::new().unwrap();

        println!("PortAudio:");
        println!("version: {}", pa.version());
        println!("version text: {:?}", pa.version_text());
        println!("host count: {}", pa.host_api_count().unwrap());

        let default_host = pa.default_host_api().unwrap();
        println!("default host: {:#?}", pa.host_api_info(default_host));

        let def_input = pa.default_input_device().unwrap();
        let input_info = pa.device_info(def_input).unwrap();
        let sample_rate = input_info.default_sample_rate;
        println!("Defaul Input Sample Rate: {}", sample_rate);
        println!("Default input device info: {:#?}", &input_info);

        let latency = input_info.default_low_input_latency;

        let input_params =
            pa::StreamParameters::<u8>::new(def_input, 1, INTERLEAVED, latency);

        pa.is_input_format_supported(input_params, sample_rate)
            .unwrap();

        let settings = pa::InputStreamSettings::new(input_params, sample_rate, FRAMES);

        let (sender, receiver) = ::std::sync::mpsc::channel();

        let _handle = std::thread::spawn(move || {
            let callback = move |pa::InputStreamCallbackArgs { buffer, frames, .. }| {
                assert!(frames == FRAMES as usize);
                sender.send(buffer.to_vec()).unwrap();
                // println!("buffer: {:?}", buffer);
                pa::Continue
            };

            let mut stream = pa.open_non_blocking_stream(settings, callback).unwrap();

            stream.start().unwrap();

            while let true = stream.is_active().unwrap() {}

            stream.stop().unwrap();
        });

        loop {
            if let Ok(data) = receiver.try_recv() {
                send_audio.send(data).unwrap();
            }
        }
    });

    let mut app = tide::with_state(Arc::new(state));

    let (playback_sender, mut playback_recv): (
        tokio::sync::mpsc::Sender<SoundPacket>,
        tokio::sync::mpsc::Receiver<SoundPacket>,
    ) = tokio::sync::mpsc::channel(64);
    let sender_clone = playback_sender.clone();

    let player_thread = tokio::spawn(async move {
        let pa = pa::PortAudio::new().unwrap();
        println!("PortAudio:");
        println!("version: {}", pa.version());
        println!("version text: {:?}", pa.version_text());
        println!("host count: {}", pa.host_api_count().unwrap());

        let default_host = pa.default_host_api().unwrap();
        println!("default host: {:#?}", pa.host_api_info(default_host));

        let def_output = pa.default_output_device().unwrap();
        let output_info = pa.device_info(def_output).unwrap();
        let sample_rate = output_info.default_sample_rate;
        println!("Default ouput device info: {:#?}", &output_info);
        println!("Default sample rate: {}", sample_rate);

        let latency = output_info.default_low_output_latency;
        println!("Output testing");

        let output_params =
            pa::StreamParameters::<u8>::new(def_output, 1, INTERLEAVED, latency);

        pa.is_output_format_supported(output_params, sample_rate)
            .unwrap();

        let settings = pa::OutputStreamSettings::<u8>::new(output_params, sample_rate, FRAMES);

        let (send_play, recv_play): (
            std::sync::mpsc::Sender<Vec<u8>>,
            std::sync::mpsc::Receiver<Vec<u8>>,
        ) = std::sync::mpsc::channel();

        let _handle = std::thread::spawn(move || {
            let callback = move |pa::OutputStreamCallbackArgs { buffer, frames, .. }| {
                assert!(frames == FRAMES as usize);
                println!("Output Buffer: {:?}", buffer);
                if let Ok(data) = recv_play.try_recv() {
                    for (output_sample, input_sample) in buffer.iter_mut().zip(data.iter()) {
                        *output_sample = *input_sample;
                    }
                }
                // println!("Output Buffer: {:?}", buffer);
                pa::Continue
            };
            let mut stream = pa.open_non_blocking_stream(settings, callback).unwrap();
            stream.start().unwrap();
            while let true = stream.is_active().unwrap() {}
            stream.stop().unwrap()
        });

        loop {
            if let Some(packet) = playback_recv.recv().await {
                println!("Audio processing");
                send_play
                    .send(packet.data)
                    .map_err(|_e| {
                        println!("Audio failed because? {}", _e);
                        tide::http::Error::from_str(
                            tide::StatusCode::BadRequest,
                            "Something happened shaa",
                        )
                    })
                    .unwrap();
            }
        }
    });

    app.at("/ws").get(WebSocket::new(
        move |req: Request<std::sync::Arc<TideState>>, wsc: WebSocketConnection| {
            let final_clone = sender_clone.clone();
            let wsc_clone = wsc.clone();
            let mut rx_audio = WatchStream::new(recv_audio.clone());
            tokio::spawn(async move {
                loop {
                    if let Some(data) = rx_audio.next().await {
                        let frame = Packet {
                            data: Some(encode(data)),
                            frame_type: None,
                            epoch_time: None,
                            encoding: None,
                            packet_type: PacketType::Audio,
                        };
                        // println!("frame: {:?}", frame);
                        let json = serde_json::to_string(&frame).unwrap();
                        wsc_clone.send(Message::text(json)).await.unwrap();
                    } else {
                        println!("Stream ended");
                        break;
                    }
                }
                // receive audio data
            });

            async move {
                println!("Web socketss {:?}", wsc);
                let mut wsc_clone = wsc.clone();

                tokio::spawn(async move {
                    while let Some(res) = wsc_clone.next().await {
                        println!("Res: {:?}", &res);
                        if let Ok(Message::Text(msg)) = res.map_err(|_e| {
                            println!("The error? {}", _e);
                            tide::http::Error::from_str(
                                tide::StatusCode::BadRequest,
                                "Error happened",
                            )
                        }) {
                            if let Ok(packet) =
                                serde_json::from_str::<SoundPacket>(&msg).map_err(|_e| {
                                    println!("Error wey happen: {}", _e);
                                    println!("Message: {:?}", &msg);
                                    tide::http::Error::from_str(
                                        tide::StatusCode::BadRequest,
                                        "Error occured",
                                    )
                                })
                            {
                                // send sound packet to playback thread
                                final_clone.send(packet).await.unwrap();
                            }
                        }
                    }
                });
                // let _ = client_connection(wsc, req, rx).await?;
                Ok(())
            }
        },
    ));

    app.listen("0.0.0.0:8000").await.unwrap();
    player_thread.await.unwrap();
    audio_thread.join().unwrap();
    Ok(())
}
```


The `audio_thread` that sends audio buffers to the client works and I hear what I am saying on the other end but I have some noisy radio-like background sound. The `player_thread` that receives audio from the client just plays siren-like noise and I don't hear anything I am saying from my client device mic. 

You also said that it could be from the `channel`, `sample rate` and `sample format`.
channel is 1 both server and client.
sample rate is 48000 both server and client.
sample format is `u8` both server and client. 

I still don't know where the issue is coming from and the more I search how audio samples work on google the more confusing it gets.

I am really sorry for asking you to explain this to me again. Thank you!

**Comment by Nathy-bajo** [comment]
In addition to what I said this is what the `audio_thread` sends to the client to play.
My client sends and receives u8 sample format so I also send and receive it on the server side.

This is what the input_buffer sends to the client and it is also what the output_buffer expects to receive so it can play a clear audio. 

```
 [125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 126, 126, 125, 125, 125, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 126, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 128, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 130, 130, 130, 130, 130, 130, 130, 130, 130, 130, 130, 130, 130, 130, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 131, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 133, 134, 134, 133, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 134, 135, 134, 135, 134, 135, 134, 135, 134, 134, 135, 135, 134, 134, 135, 135, 135, 134, 135, 134, 134, 134, 134, 135, 135, 135, 134]
```


But this is what the client sends and expects to receive so it can also play a clear audio.

```
[118, 0, 43, 0, 2, 0, 218, 255, 243, 255, 8, 0, 24, 0, 62, 0, 70, 0, 87, 0, 92, 0, 102, 0, 126, 0, 142, 0, 146, 0, 170, 0, 233, 0, 38, 1, 38, 1, 217, 0, 154, 0, 140, 0, 123, 0, 106, 0, 131, 0, 138, 0, 103, 0, 60, 0, 33, 0, 68, 0, 129, 0, 122, 0, 40, 0, 2, 0, 249, 255, 254, 255, 2, 0, 21, 0, 64, 0, 52, 0, 65, 0, 103, 0, 98, 0, 76, 0, 79, 0, 112, 0, 147, 0, 165, 0, 155, 0, 167, 0, 185, 0, 147, 0, 133, 0, 165, 0, 168, 0, 91, 0, 233, 255, 179, 255, 200, 255, 39, 0, 91, 0, 93, 0, 80, 0, 50, 0, 78, 0, 127, 0, 174, 0, 214, 0, 250, 0, 20, 1, 241, 0, 109, 0, 11, 0, 25, 0, 64, 0, 62, 0, 255, 255, 236, 255, 34, 0, 52, 0, 37, 0, 43, 0, 51, 0, 59, 0, 54, 0, 42, 0, 57, 0, 63, 0, 69, 0, 99, 0, 121, 0, 97, 0, 1, 0, 153, 255, 154, 255, 202, 255, 243, 255, 30, 0, 30, 0, 207, 255, 143, 255, 188, 255, 16, 0, 82, 0, 106, 0, 65, 0, 22, 0, 209, 255, 141, 255, 190, 255, 57, 0, 150, 0, 125, 0, 37, 0, 248, 255, 213, 255, 211, 255, 252, 255, 35, 0, 34, 0, 242, 255, 174, 255, 126, 255, 145, 255, 145, 255, 160, 255, 194, 255, 137, 255]
```


This is what my client websocket looks like.
https://github.com/Nathy-bajo/audio_client/blob/master/lib/models/sound_packet.dart

I use a `dart` package called `sound_stream` that can record and play real-time audio and send to websocket server.

How can I go about making them send the right sample please?

**Comment by pdeljanov** [comment]
According to the SoundStream docs, it only plays/records mono audio with signed 16-bit PCM samples. Even though it takes/provides these samples in a `UInt8List`, that buffer still contains samples that are 16-bit. This is a pretty common technique for audio libraries to be generic across many sample formats.

The "siren" sound is because on the Rust side you configured PortAudio for 8-bit audio. You then played each `u8` in the received buffer back directly. So in essence, you played the upper and lower bytes of each real audio sample as individual samples.

My suggestion is to use signed 16-bit PCM on both ends for playback and recording. You won't need to make any changes to the Dart code. However, on the Rust side, you'll need to convert between `Vec<i16>` and `Vec<u8>` when you send/receive your data buffers. You can use the `i16::to_ne_bytes` and `i16::from_ne_bytes` for this.

This solution will also resolve the "static" which is expected with 8-bit samples because there's not enough resolution to faithfully represent your original signal. You can Google "quantization error" to learn more, but basically you want 16-bit samples at a minimum.

**Comment by Nathy-bajo** [comment]
So the `input_params` and `output_params` should be `i16` not `u8` I should then do the conversion from `Vec<i16>`to `Vec<u8>` using `i16::to_ne_bytes` before encoding and sending it to the client and the same thing receiving the `Vec<u8>` from the client and convert it to `Vec<i16>` using `i16::from_ne_bytes` before playing the audio.

The sample rate on the server side is 48000. Should I also make the client sample rate 48000?

The frames and interleaved. Do they also match the ones on the client?

I really appreciate your help by the way!!
I have literally been spending days on the internet searching for answers and how audio samples work.

I will probably have to dive in further.

**Comment by Nathy-bajo** [comment]
> According to the SoundStream docs, it only plays/records mono audio with signed 16-bit PCM samples. Even though it takes/provides these samples in a `UInt8List`, that buffer still contains samples that are 16-bit. This is a pretty common technique for audio libraries to be generic across many sample formats.
> 
> The "siren" sound is because on the Rust side you configured PortAudio for 8-bit audio. You then played each `u8` in the received buffer back directly. So in essence, you played the upper and lower bytes of each real audio sample as individual samples.
> 
> My suggestion is to use signed 16-bit PCM on both ends for playback and recording. You won't need to make any changes to the Dart code. However, on the Rust side, you'll need to convert between `Vec<i16>` and `Vec<u8>` when you send/receive your data buffers. You can use the `i16::to_ne_bytes` and `i16::from_ne_bytes` for this.
> 
> This solution will also resolve the "static" which is expected with 8-bit samples because there's not enough resolution to faithfully represent your original signal. You can Google "quantization error" to learn more, but basically you want 16-bit samples at a minimum.

So the `input_params` and `output_params` should be `i16` not `u8` I should then do the conversion from `Vec<i16>`to `Vec<u8>` using `i16::to_ne_bytes` before encoding and sending it to the client and the same thing receiving the `Vec<u8>` from the client and convert it to `Vec<i16>` using `i16::from_ne_bytes` before playing the audio.

The sample rate on the server side is 48000. Should I also make the client sample rate 48000?

The frames and interleaved. Do they also match the ones on the client?

I really appreciate your help by the way!!
I have literally been spending days on the internet searching for answers and how audio samples work.

I will probably have to dive in further.

**Comment by pdeljanov** [comment]
> So the `input_params` and `output_params` should be `i16` not `u8` I should then do the conversion from `Vec<i16>`to `Vec<u8>` using `i16::to_ne_bytes` before encoding and sending it to the client and the same thing receiving the `Vec<u8>` from the client and convert it to `Vec<i16>` using `i16::from_ne_bytes` before playing the audio.
> 
> The sample rate on the server side is 48000. Should I also make the client sample rate 48000?
> 
> The frames and interleaved. Do they also match the ones on the client?
> 
> I really appreciate your help by the way!! I have literally been spending days on the internet searching for answers and how audio samples work.
> 
> I will probably have to dive in further.

Sounds about right. Since you're dealing with mono audio, interleaved vs. planar isn't something you need to worry about. But generally these audio input/output libraries use interleaved because it works better for streaming.

Aside from that, just always remember you need to match the three key parameters for proper playback:
1. Sample rate
2. Sample format
3. Channel count

Best of luck!

**Comment by Nathy-bajo** [comment]
I was able to fix it.

Thank you once again!!!

**Comment by Nathy-bajo** [comment]
> > So the `input_params` and `output_params` should be `i16` not `u8` I should then do the conversion from `Vec<i16>`to `Vec<u8>` using `i16::to_ne_bytes` before encoding and sending it to the client and the same thing receiving the `Vec<u8>` from the client and convert it to `Vec<i16>` using `i16::from_ne_bytes` before playing the audio.
> > The sample rate on the server side is 48000. Should I also make the client sample rate 48000?
> > The frames and interleaved. Do they also match the ones on the client?
> > I really appreciate your help by the way!! I have literally been spending days on the internet searching for answers and how audio samples work.
> > I will probably have to dive in further.
> 
> Sounds about right. Since you're dealing with mono audio, interleaved vs. planar isn't something you need to worry about. But generally these audio input/output libraries use interleaved because it works better for streaming.
> 
> Aside from that, just always remember you need to match the three key parameters for proper playback:
> 
> 1. Sample rate
> 2. Sample format
> 3. Channel count
> 
> Best of luck!

I was able to fix it.

Thank you once again!!!
